docs/v1/guidance-for-generating-p4info.md

```diff
@@ -0,0 +1,363 @@
+
+bitwidth: 10
+
+    Reason: Type T2_t is the first type name in type_list(f2), but it
+    has no p4runtime_translation on it, so even though T1_t does, that
+    is ignored.  Use the width 10 from bit<10>, the last element of
+    type_list(f2).
+```
+
+Based on the P4 code snippet above (copied below for easy reference),
+the value below starting with `type_info {` should be in the P4Info
+message describing the program, because of the `type` definitions.
+There is never anything put into a P4Info message because of `typedef`
+definitions in a P4 program.
+
+Note that the bit width of 10 appears in the P4Info message for any
+`type`s "built on top of" a `bit<10>`, _unless that type has its own
+`p4runtime_translation` annotation_.
+
+```
+typedef bit<10> T1uint_t;
+@p4runtime_translation("mycompany.com/psa/v1/T1_t", 32)
+type T1uint_t T1_t;
+type T1_t T2_t;
+T2_t f2;
+```
+
+```
+type_info {
+  new_types {
+    key: "T2_t"
+    value {
+      original_type {
+        bitstring {
+          bit {
+            bitwidth: 10
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+Type `T2_t` is described via an `original_type` message, because the
+definition of `T2_t` in the program, `type T1_t T2_t`, does not have a
+`p4runtime_translation` annotation.
+
+Based on the P4 code snippet above, there is no reason to include a
+description for the type `T1_t` in the P4Info message.  If there were
+some other variable other than `f2` in the program declared with type
+`T1_t`, and that variable caused the type name `T1_t` to be included
+in the P4Info message, then the following description for type `T1_t`
+must be included in the `type_info` field.
+
+```
+type_info {
+  new_types {
+    key: "T1_t"
+    value {
+      translated_type {
+        uri: "mycompany.com/psa/v1/T1_t"
+        sdn_bitwidth: 32
+      }
+    }
+  }
+}
+```
+
+Type `T1_t` is described via a `translated_type` message, because the
+definition of `T1_t` in the program, `type T1uint T1_t`, has a
+`p4runtime_translation` annotation.
+
+
+## Example 3: field with user-defined `type` with `p4runtime_translation` annotation
+
+It is not clear whether there are strong use cases for declaring a
+`type` based upon another `type` in a P4_16 program.  However,
+currently the P4_16 language and `p4c` compiler allow it, so it seems
+to be a good idea to have predictable rules to follow for what the
+P4Info message contents should be, and how the resulting system should
+behave.
+
+The basic design described here tries to keep things fairly
+straightforward to explain and understand, if a P4_16 program does so.
+
+If a constrained value `f` has a `type` name as the first element of
+the list `type_list(f)`, then that type name is used for `f`,
+regardless of whether it has a `p4runtime_translation` annotation.
+
+In the absence of such an annotation on that `type`, no P4runtime
+translation is done for that type, _even if a later type in
+`type_list(x)` does have such an annotation_.
```

Comment:
Doesn't this text apply to the previous example (example 2)?

Reply:
It does.  I moved that text to the previous example, which has been renumbered example 3 to make room for a new simpler example 2, an example that is the simple one you suggested.